### PR TITLE
add slabs

### PR DESCRIPTION
**Pull request by plumpudding**
see #1379 

could defenitely be done better, but works perfectly fine like this
I tried to keep this compliant to your syntax, shouldn't be too much to fix
do you want this for rv2 too?

### CODE REVIEW

**Comment by TheJulianJES** [comment]
Please use the right format, sometimes you used (something) and not ( something )

File: src/main/java/appeng/block/AEBaseItemBlockSlab.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package appeng.block;
+
+import net.minecraft.block.Block;
+import net.minecraft.item.ItemSlab;
+
+public class AEBaseItemBlockSlab extends ItemSlab {
+
+	public AEBaseItemBlockSlab(Block b, AEBaseSlabBlock sb, AEBaseSlabBlock db, Boolean id) {
+		super( b, sb, db, id );
+		System.out.println(b);
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseItemBlockSlab.java]: Remove these.

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	public void registerBlockIcons(IIconRegister reg) { }
+
+	@Override
+	public Item getItemDropped(int meta, Random rand, int fortune)
+	{
+	    return Item.getItemFromBlock(this.slabs);
+	}
+
+	@Override
+	public int quantityDropped(int meta, int fortune, Random random)
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Why override it?

**Comment by plumpudding** [comment]
@yueh which exactly do you mean?

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	{
+	    return this.field_150004_a ? 2 : 1;
+	}
+
+	@Override
+	public ItemStack getPickBlock(MovingObjectPosition target, World world, int x, int y, int z)
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Do not implement deprecated API, use the alternative.

**Comment by plumpudding** [comment]
more specific? which of them is deprecated?

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        this.setStepSound( block.stepSound );
+        this.useNeighborBrightness = true;
+		if (!isDoubleSlab)
+		{
+			this.dSlabs = new AEBaseSlabBlock( block, meta, features, true, name + ".double" ).setSlabs(this);
+			this.slabs = this;
+		} else {
+			this.dSlabs = this;
+		}
+		this.features = isDoubleSlab ? new SlabBlockFeatureHandler( features, this ) : new SlabBlockFeatureHandler( features, this );
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: What is the difference here?

**Comment by plumpudding** [comment]
-.- tabs and spaces and eclipse

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+	public AEBaseSlabBlock( AEBaseBlock block, int meta, EnumSet<AEFeature> features, boolean isDoubleSlab, String name )
+	{
+        super( isDoubleSlab, block.getMaterial() );
+		this.block = block;
+		this.meta = meta;
+		this.name = name;
+		this.isDoubleSlab = isDoubleSlab;
+		this.setBlockName( "appliedenergistics2." + name );
+        this.setHardness( block.getBlockHardness( null, 0, 0, 0 ) );
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Why call the getters, if you do not change anything? Same for the `setResistance`.

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        this.setResistance( block.getExplosionResistance( null ) * 5.0F / 3.0F );
+        this.setStepSound( block.stepSound );
+        this.useNeighborBrightness = true;
+		if (!isDoubleSlab)
+		{
+			this.dSlabs = new AEBaseSlabBlock( block, meta, features, true, name + ".double" ).setSlabs(this);
+			this.slabs = this;
```
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Why reassign `this` multiple times?

**Comment by plumpudding** [comment]
will fix all this later today or tomorrow

now have to do homework -.-

File: src/main/java/appeng/block/AEBaseSlabBlock.java (hunk 2 above)
Comment (yueh) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: It does exactly the same as `BlockSlab.quantityDropped()`.

And use the "Add a line note" button, it is easier to keep the context.

**Comment by TheJulianJES** [comment]
@plumpudding lol

**Comment by plumpudding** [comment]
not funny

**Comment by yueh** [comment]
Line notes are always related to the last displayed line. No idea why github does not highlight them and also show a snippet of the code after it.

**Comment by TheJulianJES** [comment]
@plumpudding O_o I know, good thing I don't have to deal with that anymore :)

**Comment by plumpudding** [comment]
Is that better? @yueh

File: src/main/java/appeng/block/AEBaseSlabBlock.java (hunk 5 above)
Comment (plumpudding) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: blockhardness and resistance are private in the block. I have to use the getter.

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		this.meta = meta;
+		this.name = name;
+		this.isDoubleSlab = isDoubleSlab;
+		this.setBlockName( "appliedenergistics2." + name );
+		this.setHardness( block.getBlockHardness( null, 0, 0, 0 ) );
+		this.setResistance( block.getExplosionResistance( null ) * 5.0F / 3.0F );
+		this.setStepSound( block.stepSound );
+		this.useNeighborBrightness = true;
+		if (!isDoubleSlab) this.dSlabs = new AEBaseSlabBlock( block, meta, features, true, name + ".double" ).setSlabs(this);
+		this.features = !isDoubleSlab ? new SlabBlockFeatureHandler( features, this ) : null;
```
Comment (plumpudding) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Double slabs don't need a handler. They are handled by their non-double equivalent's handler.

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	public void registerBlockIcons(IIconRegister reg) { }
+
+	@Override
+	public Item getItemDropped(int meta, Random rand, int fortune)
+	{
+		return Item.getItemFromBlock(this.slabs);
+	}
+
+	@Override
+	public ItemStack getPickBlock(MovingObjectPosition target, World world, int x, int y, int z)
```
Comment (plumpudding) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: The non-deprecated API doesn't work.

File: src/main/java/appeng/block/AEBaseSlabBlock.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+	public AEBaseSlabBlock( AEBaseBlock block, int meta, EnumSet<AEFeature> features, boolean isDoubleSlab, String name )
+	{
+		super( isDoubleSlab, block.getMaterial() );
+		this.block = block;
+		this.meta = meta;
+		this.name = name;
+		this.isDoubleSlab = isDoubleSlab;
+		this.setBlockName( "appliedenergistics2." + name );
+		this.setHardness( block.getBlockHardness( null, 0, 0, 0 ) );
```
Comment (plumpudding) [inline src/main/java/appeng/block/AEBaseSlabBlock.java]: Hardness and Resistance are private in the block class. I have to use the getter.

**Comment by plumpudding** [comment]
any updates on the review of this?

**Comment by thatsIch** [comment]
It is probably better for scaling reason to use 2 different blocks for double and single slab.

The rest looks fine to me after a squash.

**Comment by thatsIch** [comment]
Rebase required

**Comment by plumpudding** [comment]
I am unable to seperate it into 2 blocks. I can push the broken version to a seperate branch of my fork so you can clone it and work on it if you want.

**Comment by thatsIch** [comment]
what restriction are you reaching?

**Comment by plumpudding** [comment]
I have to access the double slab to register the single slab and the other way arround. So I have to get the other block from the API, but I always get Nullpointers because they are not reachable through the API at that point. It's been some time since I've written that so this is just roughly what the problem was.

**Comment by thatsIch** [comment]
I just did a quick refactor to see the problem. I noticed the stopper, else it would have a cyclic dependency due to the nature of `IAEFeature`

Then a working implementation would suffice for me. We can probably fix this later, if `IAEFeature`s are gonna reworked.

**Comment by plumpudding** [comment]
So tomorrow I'll do some more testing and then squash. Would that be OK then?

**Comment by thatsIch** [comment]
Sure, take your time.

**Comment by yueh** [comment]
The main issues is probably abusing the `AEBaseSlabBlock ` as collection, block and `IAEFeature`.
It might be a good example for moving the `IAEFeature` into its own class, but probably something we should change in general when reworking them and also structuring the packages in a different way.

**Comment by plumpudding** [comment]
I had to recreate the master. See #1500.

**Pull request closed without merging**